Fix macOS SDK builds without ldap_init_fd

tiran opened the pull request:
macOS system libldap 2.4.28 does not have ldap_init_fd symbol. Disable
initialize_fd when Apple libldap 2.4.28 is detected.

Also run some macOS tests on Travis CI. Since the SDK does not ship
slapd, testing is rather limited.

Fixes: https://github.com/python-ldap/python-ldap/issues/359
Signed-off-by: Christian Heimes <cheimes@redhat.com>

encukou commented:
Could you also document when the functionality is available?

tiran commented:
> Could you also document when the functionality is available?

I would liek to but I don't know. OpenLDAP 2.4.28 has ``ldap_init_fd``. It was introduced around 2.4.0. I have no clue why the symbol is missing in the Apple builds. It could be patched out or hidden by a downstream patch.

encukou commented:
From the patch, I gather that on a Mac, you won't get fileno support even if you compile OpenLDAP 2.4.28 yourself.

tiran commented:
> From the patch, I gather that on a Mac, you won't get fileno support even if you compile OpenLDAP 2.4.28 yourself.

2.4.28 was released over 8 years ago. I don't we have to care about ancient versions except for Apple system builds.

I'm also open to better ways to detect system libldap on macOS. I don't have access to macOS to investigate the issue.

encukou commented:
I'm OK with the patch itself. I just think it needs a note in the documentation that says when the feature is available. From the current text, it seems that *fileno* is always usable.

Also, you've now rebased onto master, which I don't understand. Shouldn't this go into 3.2.1?

tiran commented:
> I'm OK with the patch itself. I just think it needs a note in the documentation that says when the feature is available. From the current text, it seems that _fileno_ is always usable.

I have updated documentation to mention ``INIT_FD_AVAIL`` and the macOS issue.

> Also, you've now rebased onto master, which I don't understand. Shouldn't this go into 3.3.1?

The fix should land in master and 3.3.1. New features land in master and then are back-ported to bug fix branches.

tiran commented:
Please merge #361 and #362 before this PR.

encukou commented:
> New features land in master and then are back-ported to bug fix branches.

That way makes it harder to merge. It is the CPython workflow, but this repo doesn't have Miss Islington like CPython.

tiran commented:
> That way makes it harder to merge. It is the CPython workflow, but this repo doesn't have Miss Islington like CPython.

I use the same workflow in all projects. Patches land in master first and then get backported to maintenance branches. The patch would have conflicted either way.

To make it easier for you I have created a backport branch, https://github.com/python-ldap/python-ldap/pull/363

encukou commented on Doc/reference/ldap.rst:
```diff
@@ -49,6 +49,8 @@ This module defines the following functions:
    and explicitly closed after the :class:`~ldap.ldapobject.LDAPObject` is
    unbound. The internal connection type is determined from the URI, ``TCP``
    for ``ldap://`` / ``ldaps://``, ``IPC`` (``AF_UNIX``) for ``ldapi://``.
+   The is not available on macOS when python-ldap is compiled with system
```
```suggestion
   The parameter is not available on macOS when python-ldap is compiled with system
```
